### PR TITLE
fix(model_runner): fail-fast on parameters left uninitialized by checkpoint (fixes #26745)

### LINKED ISSUE #26745: [Bug] Silent NaN corruption when serving VLM model without visual encoder weights
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Bug Description

When serving a VLM model (e.g., Qwen3.5-122B-A10B) with a checkpoint that's missing visual encoder weights (e.g., after Megatron training that only trains the language model), SGLang silently serves NaN/garbage output for all chat requests. No error or warning is raised at any point.

### Root Cause
3 issues combine:

1. **No weight loading validation:** `load_weights()` doesn't check or report which parameters remain uninitialized. Visual encoder params are allocated with `torch.empty()` and never loaded from checkpoint — they contain arbitrary memory. No warning is emitted.

2. **Warmup sends image request to uninitialized visual encoder:** `_execute_server_warmup()` detects VLM architecture and sends a 32×32 PNG. The uninitialized visual encoder produces NaN embeddings. GDN (recurrent) layers propagate NaN to subsequent positions. Full attention layers write NaN K,V to the KV pool. After warmup completes, these NaN KV slots are cached in the radix tree. Subsequent chat requests that share a prefix with warmup (`<|im_start|>user\n`) pick up these corrupted slots via `cache_unfinished_req`'s prefix slot reuse — even though the scheduler correctly rejected prefix sharing (`prefix_len=0`) and extend computed correct KV with fresh slots. Once corrupted KV enters the cache, it silently propagates to all subsequent requests sharing that prefix.

### Impact
- ALL /v1/chat/completions requests produce garbage after warmup
- /v1/completions (raw tokens, no chat template) is unaffected (no shared prefix)
- No error/warning in logs caused silent failure

#### 1. Weight loading validation (aligns with RFC #24703)

 After load_weights(), compare loaded keys against model.state_dict().keys(). Warn or error on unloaded parameters:
 ```
  loaded_keys = model.load_weights(weights)
  all_keys = set(model.state_dict().keys())
  unloaded = all_keys - loaded_keys
  if unloaded:
      raise ValueError(f"Weights not initialized from checkpoint: {unloaded}")
 ```

#### 2. Warmup should validate visual encoder state

 Before sending image warmup, check if visual weights are actually loaded:
 ```
  if hasattr(model, 'visual') and model.visual is not None:
      first_param = next(model.visual.parameters())
      if first_param.isnan().any() or first_param.abs().max() > 1e30:
          # Skip image warmup, use text-only
          pass
 ```

#### 3. A general "skip vision encoder" mode

 We note that SGLang's --language-only flag is specifically designed for encoder-decoder prefill disaggregation (EPD), not a general "skip  vision encoder" mode (see #21140 > "language_only requires --encoder-urls and is restricted to Qwen architectures only").

 vLLM's --language-model-only covers this use case cleanly. Would SGLang consider a similar general-purpose flag (separate from EPD's --language-only) that:
 - Skips visual module allocation (or stubs it out)
 - Skips image warmup
 - Rejects multimodal requests at the API level

If the current architecture doesn't support this cleanly, fixes #1 and #2 would already prevent the silent failure

### Related Issues
  - #24703 — RFC: Refactor weight loading code (proposes set[str] return from load_weights)
  - #22701 — Corrupted weights, missing quant init for multimodal models
  - #20043 — NaN in hidden states with modelopt_fp4 (--enable-nan-detection)
  - #22935 — Mamba radix cache: fresh prefill prefix hits become 0-hit after split
  - #21140 — Clarification that --language-only is for EPD, not general "skip vision encoder"

### Reproduction

  1. Serve a VLM model with checkpoint missing visual weights
  2. Server warmup fires automatically (sends image)
  3. Send a chat request: curl /v1/chat/completions -d '{"messages":[{"role":"user","content":"What is 2+2?"}]}'
  4. Observe garbage output > no error in logs

### Environment

- Model: Qwen3.5-122B-A10B (hybrid GDN + attention, VLM)
- Checkpoint: Megatron-converted, 333 visual encoder params missing

python3 -m sglang.check_env
```
Python: 3.12.3
CUDA available: True
GPU 0-7: NVIDIA H200 (Compute Capability 9.0)
CUDA: 12.9 / Driver: 580.126.09
PyTorch: 2.9.1+cu129
sglang: 0.5.9
flashinfer: 0.6.3
triton: 3.5.1
transformers: 5.2.0
Topology: 8x H200, NV18 interconnect
```

### PR DESCRIPTION
**Pull request by abinggo**
## Summary

Reporter (#26745) traced a 3-step silent failure on VLMs whose checkpoint is missing the visual encoder weights:

1. `model.load_weights(weights)` does not check or report which parameters remain uninitialized — the visual encoder slots keep whatever bit pattern `torch.empty()` allocated.
2. The HTTP server's `_execute_server_warmup()` autodetects a VLM architecture and sends a 32×32 PNG, which produces NaN embeddings; GDN recurrent layers propagate NaN, full-attention layers write NaN K, V to the KV pool.
3. Those NaN KV slots get cached in the radix tree, so every subsequent `/v1/chat/completions` request sharing the warmup prefix returns garbage with **no error in the logs**.

This patch implements fix #1 from the issue (fail-fast at load), which makes fix #2 (warmup-time guard) redundant — the server never reaches warmup with poisoned weights.

Fixes #26745.

## Approach

A heuristic detector for the `torch.empty()` byte-pattern signature on floating dtypes:

- **`_has_uninitialized_signature(tensor)`** — flags floating tensors that contain NaN, Inf, or any element with abs magnitude > `_UNINITIALIZED_MAGNITUDE_THRESHOLD = 1e30`. Inspection is bounded to the first `_UNINITIALIZED_SAMPLE_BUDGET = 256` elements (uninitialized memory is dense — NaN/Inf occupy ~50% of decoded float bytes — so a few hundred samples catch the failure with overwhelming probability while keeping per-tensor cost bounded on 100B+ parameter models).
- **`_validate_loaded_weights(model, max_report=10) -> (sample, total)`** — walks `named_modules()` and inspects every module's direct floating params via `named_parameters(recurse=False)`. Non-leaf modules that own their own `torch.empty` parameters (e.g. decoder layers that mix child sub-modules with directly owned expert weights) are still inspected; their children's params are not re-inspected at the parent level. Skips `meta` tensors and non-floating dtypes (quantized int4/int8 weights). Returns `(sample, total)` so the error message can say "at least N" rather than truncating silently.
- **Call site** — `ModelRunner.load_model()` calls the validator immediately after `self.loader.load_model(...)`, **before** downstream consumers (FP8 KV-cache scale loading, sliding-window probing, draft-model setup, runtime quant post-processing) can surface their own opaque error against an already-poisoned model.

Disable via `SGLANG_SKIP_WEIGHT_VALIDATION=1` (registered in `environ.py` under "Model & File Download"). Only enable when knowingly serving a partial checkpoint — e.g. RL warm-up where weights stream in later.

## On heuristic vs precise contract

This is intentionally a heuristic. A precise contract via `load_weights() -> set[str]` is RFC #24703; this defense remains useful as belt-and-suspenders even after that lands, because:
- It catches the "checkpoint dropped a whole submodule" failure mode without depending on every model parser updating its return type.
- It detects bit-pattern garbage from non-checkpoint sources (e.g. corrupted reads, partial DMA failures) that key tracking would not catch.

## Test plan

`test/registered/model_loading/test_validate_loaded_weights.py` (`register_cpu_ci(est_time=10, suite="base-a-test-cpu")`):

**`_has_uninitialized_signature` (12 cases)**
- zeros / randn → clean
- NaN / Inf / -Inf / 1e35 in prefix → dirty
- threshold boundary just-over (1.5e30) / just-under (1e29) → dirty / clean
- empty tensor → clean
- NaN past sample budget → not inspected (bounded-sampling contract pin)
- bf16 + fp16 dtype variants

**`_validate_loaded_weights` (10 cases)**
- clean model returns `([], 0)`
- visual tower uninitialized: weight + bias both flagged (recurse=False per-module catches secondary params)
- both towers uninitialized: total ≥ 4
- max_report=10 over 30 dirty params: `len(sample) == 10`, `total == 30`
- meta-tensor skip
- integer-only module skip
- secondary param caught (clean weight + dirty bias on same leaf)
- non-leaf direct param caught (DeepseekV4-style mixed container)
- bf16 dtype
- `requires_grad=False` floating param

```
pre-commit run --files \
    python/sglang/srt/environ.py \
    python/sglang/srt/model_executor/model_runner.py \
    test/registered/model_loading/test_validate_loaded_weights.py
ruff..............................Passed
black.............................Passed
isort.............................Passed
codespell.........................Passed
check_registered_tests............Passed
spdx..............................Passed
```

AI assistance was used in drafting this fix and tests; the human contributor reviewed every line and ran the test plan locally.

---
### CI States

Latest PR Test (Base): :x: [Run #26680411130](https://github.com/sgl-project/sglang/actions/runs/26680411130)
Latest PR Test (Extra): :x: [Run #26680411086](https://github.com/sgl-project/sglang/actions/runs/26680411086)

### CODE REVIEW

**Comment by abinggo** [comment]
Thanks for the careful read, @gemini-code-assist. Both points landed well — pushed `c35eabb2f7 → 6b0b9f7bea`.

**Per-tensor (`_has_uninitialized_signature`):** factored the fused predicate into a tiny helper `_bytes_look_uninitialized(sample) -> torch.Tensor` and reduced the public function to a single `(...).any().item()`. `~torch.isfinite(x) | (x.abs() > THRESH)` covers NaN, ±Inf, and the magnitude check in one elementwise pass.

**Module walk (`_validate_loaded_weights`):** rewrote into the two-pass batched shape you suggested.

* Pass 1 collects bounded prefix samples in `named_modules()` walk order (kept stable so the truncated `sample` list still surfaces top-of-tree culprits first when the cap fires).
* Pass 2 groups samples by `(device, dtype)` so `torch.cat` is a copy with no implicit cast / H2D transfer, then runs the predicate once per group. The 99.9% common case (no uninitialized params) collapses to one `.any().item()` per `(device, dtype)` tuple — typically one or two — regardless of model size.
* Slow path runs only when a group flips: each entry inside that group gets the single-sync per-prefix check to identify exact culprits, so the failure-mode error message keeps its precise list.

`(sample, total)` API is unchanged; existing tests pass byte-for-byte. Smoke-tested against the same 21 cases in `test_validate_loaded_weights.py` plus a few extra mixed-dtype permutations to confirm the group key handles bf16/fp16/fp32 in the same model. CI's `base-a-test-cpu` suite will run them when picked up.
